### PR TITLE
 support generate document id with redirections

### PR DESCRIPTION
**Pull request by OsmondJiang**

### CODE REVIEW

**Comment by OsmondJiang** [comment]
#2821

File: docs/specs/metadata.yml [hunk 1]
```diff
@@ -36,3 +36,66 @@ outputs:
...
+      siteBasePath: sccm
+    name: sccm
+    product: Win
+    content: "**/*"
+    redirections:
+      sccm/mdm/index.md: /sccm/mdm/understand/hybrid-mobile-device-management
+  sccm/mdm/understand/hybrid-mobile-device-management.md:
+outputs:
+  sccm/mdm/understand/hybrid-mobile-device-management.json: |
+    { "content": "", "id": "1c42a151-6248-df0c-15e6-3b1158d7bee3", "versionIndependentId": "beb14a44-b076-14b8-b9c5-b42f9e36b973" }
```
Comment (yufeih) [inline docs/specs/metadata.yml]: nit: you can just check `id` and `versionIndependentId` here

File: src/docfx/build/markdown/BuildMarkdown.cs [hunk 2]
```diff
@@ -37,15 +40,40 @@ internal static class BuildMarkdown
...
 
             // TODO: make build pure by not output using `context.Report/Write/Copy` here
-            context.Report(file, markup.Errors);
+            var errors = error == null ? markup.Errors : markup.Errors.Concat(new[] { error });
+            context.Report(file, errors);
             context.WriteJson(model, file.OutputPath);
 
             return Task.FromResult(dependencyMapBuilder.Build());
+
+            (string id, string versionIndependentId, DocfxException error) GetIds()
```
Comment (yufeih) [inline src/docfx/build/markdown/BuildMarkdown.cs]: put this to `RedirectionMap` to share it with `BuildSchemaDocument`

File: docs/specs/config.yml [hunk 3]
```diff
@@ -241,4 +241,15 @@ inputs:
...
   build.log: |
-    ["error","invalid-redirection","The docs/toc.md shouldn't belong to redirections since it's a TableOfContents"]
+    ["error","invalid-redirection","The 'docs/toc.md' shouldn't belong to redirections since it's a TableOfContents"]
+---
+# Conflicted redirection file
+inputs:
+  docfx.yml: |
+    redirections:
+      docs/a.md: /absolute/path1
+    redirectionsWithoutDocumentId:
```
Comment (yufeih) [inline docs/specs/config.yml]: Maybe we could **prefix** the URL with `!` to mean that it should also redirect document id?

File: src/docfx/build/markdown/BuildMarkdown.cs (hunk 2 above)
Comment (yufeih) [reply src/docfx/build/markdown/BuildMarkdown.cs]: Convention from golang, put error as the first argument, to make errors significant

File: docs/specs/config.yml (hunk 3 above)
Comment (OsmondJiang) [reply docs/specs/config.yml]: !/absolute/path1? not really easy to understand

---
In reply to: [196650230](https://github.com/dotnet/docfx/pull/2904#discussion_r196650230) [](ancestors = 196650230)

File: docs/specs/metadata.yml [hunk 4]
```diff
@@ -36,3 +36,66 @@ outputs:
...
+      sccm/mdm/index.md: /sccm/mdm/understand/hybrid-mobile-device-management
+      sccm/mdm/index2.md: /sccm/mdm/understand/hybrid-mobile-device-management
+  sccm/mdm/understand/hybrid-mobile-device-management.md:
+outputs:
+  sccm/mdm/understand/hybrid-mobile-device-management.json: |
+    { "id": "1c42a151-6248-df0c-15e6-3b1158d7bee3", "versionIndependentId": "beb14a44-b076-14b8-b9c5-b42f9e36b973" }
+  sccm/mdm/index.json: |
+    { "redirectionUrl":"/sccm/mdm/understand/hybrid-mobile-device-management", "id": "1c42a151-6248-df0c-15e6-3b1158d7bee3", "versionIndependentId": "beb14a44-b076-14b8-b9c5-b42f9e36b973" }
+  sccm/mdm/index2.json: |
+    { "redirectionUrl":"/sccm/mdm/understand/hybrid-mobile-device-management", "id": "573a60b3-e1f9-c2a1-6d92-47f36b55d89b", "versionIndependentId": "b0cc90bf-c553-6b04-e613-57559503e93c" }
```
Comment (superyyrrzz) [inline docs/specs/metadata.yml]: why there id is different?

**Pull request merged**